File: python/src/causentra/propagation.py
"""Strict W3C Trace Context injection and extraction."""

from __future__ import annotations

import re
from collections.abc import Mapping

from .types import TraceContext

_TRACEPARENT = re.compile(
    r"^00-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$",
    re.IGNORECASE,
)


def inject_trace_context(
    context: TraceContext,
    carrier: Mapping[str, str] | None = None,
    *,
    sampled: bool = True,
) -> dict[str, str]:
    """Return a copy of a carrier containing a canonical W3C traceparent."""

    result = dict(carrier or {})
    flags = "01" if sampled else "00"
    result["traceparent"] = f"00-{context.trace_id}-{context.span_id[-16:]}-{flags}"
    return result


def extract_trace_context(carrier: Mapping[str, str]) -> TraceContext | None:
    """Parse a supported W3C carrier, rejecting invalid and all-zero IDs."""

    value = next((v for k, v in carrier.items() if k.lower() == "traceparent"), None)
    if value is None:
        return None
    match = _TRACEPARENT.fullmatch(value.strip())
    if match is None:
        return None
    trace_id, span_id, _flags = (item.lower() for item in match.groups())
    if trace_id == "0" * 32 or span_id == "0" * 16:
        return None
    return TraceContext(trace_id=trace_id, span_id=span_id)

File: packages/sdk/src/propagation.ts
import type { TraceContext } from "./types.js";

const TRACE_ID_PATTERN = /^[0-9a-f]{32}$/u;
const SPAN_ID_PATTERN = /^[0-9a-f]{16}$/u;
const TRACEPARENT_PATTERN = /^00-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})$/u;

/** String carrier suitable for HTTP, messaging, or job metadata. */
export type TraceCarrier = Readonly<Record<string, string | undefined>>;

/** Valid remote W3C parent extracted from a carrier. */
export interface PropagatedTraceContext {
  readonly traceId: string;
  readonly spanId: string;
  readonly traceFlags: number;
  readonly traceState?: string;
}

/** Injects a W3C `traceparent` and optional `tracestate` without mutating input. */
export function injectTraceContext(
  context: TraceContext,
  carrier: TraceCarrier = {},
  options: { readonly sampled?: boolean; readonly traceState?: string } = {},
): Record<string, string> {
  const spanId = context.spanId.slice(0, 16);
  if (!validTraceId(context.traceId) || !validSpanId(spanId)) {
    throw new TypeError("trace context contains invalid W3C identifiers");
  }
  const result = Object.fromEntries(
    Object.entries(carrier).filter((entry): entry is [string, string] => entry[1] !== undefined),
  );
  result.traceparent = `00-${context.traceId}-${spanId}-${options.sampled === false ? "00" : "01"}`;
  if (options.traceState !== undefined) result.tracestate = options.traceState;
  return result;
}

/** Extracts a strict W3C version-00 parent; malformed/all-zero values are ignored. */
export function extractTraceContext(
  carrier: TraceCarrier,
): PropagatedTraceContext | undefined {
  const traceparent = header(carrier, "traceparent")?.trim();
  if (traceparent === undefined) return undefined;
  const match = TRACEPARENT_PATTERN.exec(traceparent);
  if (match === null) return undefined;
  const [, traceId, spanId, flags] = match;
  if (traceId === undefined || spanId === undefined || flags === undefined) return undefined;
  if (!validTraceId(traceId) || !validSpanId(spanId)) return undefined;
  const traceFlags = Number.parseInt(flags, 16);
  return {
    traceId,
    spanId,
    traceFlags,
    ...(header(carrier, "tracestate") === undefined
      ? {}
      : { traceState: header(carrier, "tracestate") as string }),
  };
}

function header(carrier: TraceCarrier, name: string): string | undefined {
  const entry = Object.entries(carrier).find(([key]) => key.toLowerCase() === name);
  return entry?.[1];
}

function validTraceId(value: string): boolean {
  return TRACE_ID_PATTERN.test(value) && !/^0+$/u.test(value);
}

function validSpanId(value: string): boolean {
  return SPAN_ID_PATTERN.test(value) && !/^0+$/u.test(value);
}
